
# HG changeset patch
# User Ping Chen <remotenonsense@gmail.com>
# Date 1658392769 -7200
# Node ID be9500248a41be9bd0ce29ea9b635886e1cf6aef
# Parent  aa857236d65fbf211185a4734ab2b2edc2110840
Bug 1778924 - Put getIncomingServer in try..catch block to fix loading nntp article. r=mkmelin


Fix the case when two servers have the same hostname.

Differential Revision: https://phabricator.services.mozilla.com/D151893

diff --git a/mailnews/news/src/NntpMessageService.jsm b/mailnews/news/src/NntpMessageService.jsm
--- a/mailnews/news/src/NntpMessageService.jsm
+++ b/mailnews/news/src/NntpMessageService.jsm
@@ -101,17 +101,16 @@ class BaseMessageService {
     messageUri,
     file,
     addDummyEnvelope,
     urlListener,
     outUrl,
     canonicalLineEnding,
     msgWindow
   ) {
-    this._logger.debug("SaveMessageToDisk", messageUri);
     let url = this.getUrlForUri(messageUri, msgWindow);
     if (urlListener) {
       url.RegisterListener(urlListener);
     }
     url.newsAction = Ci.nsINntpUrl.ActionSaveMessageToDisk;
     url.AddDummyEnvelope = addDummyEnvelope;
     url.canonicalLineEnding = canonicalLineEnding;
 
diff --git a/mailnews/news/src/NntpUtils.jsm b/mailnews/news/src/NntpUtils.jsm
--- a/mailnews/news/src/NntpUtils.jsm
+++ b/mailnews/news/src/NntpUtils.jsm
@@ -38,16 +38,22 @@ var NntpUtils = {
       keySet.add(name.split(".")[0]);
     }
 
     // Find the NNTP server that matches the hostname.
     for (let key of keySet) {
       let type = branch.getCharPref(`${key}.type`, "");
       let hostnameValue = branch.getCharPref(`${key}.hostname`, "");
       if (type == "nntp" && hostnameValue == hostname) {
-        return MailServices.accounts
-          .getIncomingServer(key)
-          .QueryInterface(Ci.nsINntpIncomingServer);
+        try {
+          return MailServices.accounts
+            .getIncomingServer(key)
+            .QueryInterface(Ci.nsINntpIncomingServer);
+        } catch (e) {
+          // In some profiles, two servers have the same hostname, but only one
+          // can be loaded into AccountManager. Catch the error here and the
+          // already loaded server will be found.
+        }
       }
     }
     return null;
   },
 };
